config/build.ts

```diff
@@ -1,33 +1,47 @@
 import { $ } from "zx";
-import { join } from "node:path";
+import { join, resolve } from "node:path";
 import { parseArgs } from "node:util";
 
+import { compileTs } from "./compileTs.ts";
 import { inlineInheritDoc } from "./inlineInheritDoc.ts";
 import { updateVersion, verifyVersion } from "./version.ts";
 import { processInvariants } from "./processInvariants.ts";
-import { rewriteSourceMaps } from "./rewriteSourceMaps.ts";
-import { rollup } from "./rollup.ts";
 import { prepareDist } from "./prepareDist.ts";
 import { postprocessDist } from "./postprocessDist.ts";
+import { verifySourceMaps } from "./verifySourceMaps.ts";
 import { prepareChangesetsRelease } from "./prepareChangesetsRelease.ts";
 
+export interface BuildStepOptions {
+  type: "esm" | "cjs";
+  rootDir: string;
+  /** build target directory, relative to `rootDir` */
+  targetDir: string;
+  jsExt: "js" | "cjs";
+  tsExt: "ts" | "cts";
+}
+export type BuildStep = {
+  (options: BuildStepOptions): void | Promise<void>;
+  // some build steps only need to run once as they don't need to be run for each build type
+  runOnce?: "leading" | "trailing";
```

Comment:
I think I'm a bit confused on the `leading`/`trailing` names here. Is this just to limit the build to CJS or ESM only? Thats the way I'm interpreting it while reading the code.

If that is indeed the case, perhaps name it more in line with that?

```suggestion
  target?: "esm" | "cjs";
```

Thoughts?

Reply:
The intention here is not really to run with the "esm" build or the "cjs" build, but with the first build or the last build.

While working on this, I'm regularly commenting out either the ESM or CJS build to iterate faster with the build that I'm working on, but in that case, I'm only left with one build, and a `target` build might get skipped.
At the same time, with only one build, both "leading" and "trailing" will still be running - they'd just both run on the same build.

Also, I'm not really sure if we will stay with these two builds - there is a good chance we'll end up with some more, depending on how we do the condition targeting (`devlopment`, `production`, `react-server` etc). 
I'll try to stay with two builds, but I really can't say right now.

The idea here is more: this build step doesn't care if we're building ESM or CJS or whatever - it's part of the package preparation, and not of a specific target. It only needs to run once, running it multiple times would be a waste.
Some of those preparation steps should be run very early in the process, some should run very late.

Reply:
That said, I'm pretty sure this might change a bit with the next PR :)